spec/GraphQLOverHTTP.md

```diff
@@ -201,8 +201,35 @@ may not be the GraphQL server but instead some intermediary such as API gateways
 
 ## Status Codes
 
-If the `data` entry in the response has any value other than `null` (when the operation has successfully executed
-without error) then the response should use the 200 (OK) status code.
-If the operation failed before or during execution, due to a syntax error, missing information, validation error
-or any other reason, then response should use the 4XX or 5XX status codes.
-It is recommended to use the same error codes as the [reference implementation](https://github.com/graphql/express-graphql).
+If the response has Content-Type GraphQL and contains a non-null `data` entry,
+then it MUST have status code `2xx`, and it SHOULD have status code `200`.
+
+If the response has Content-Type GraphQL and has a non-`2xx` status code,
+the `data` entry must be either:
+- equal to `null`
+- not present
+
+Note: The result of executing a GraphQL operation may contain partial data as well as encountered errors.
+Errors that happen during execution of the GraphQL operation typically become part of the result,
+as long as the server is still able to produce a well-formed response.
+
+In case of errors that completely prevent the successful execution of the request,
+the server SHOULD respond with the appropriate status code depending on the concrete error condition.
+
+There are four types of validation errors:
+
+1. Unparseable or invalid request body, for example `NONSENSE` or `{"qeury": "{__typename}"}`
+2. GraphQL Specification document validation errors (valid syntax, no loops in fragments, etc)
+3. Custom document validations that can be performed before execution (e.g. graphql-depth-limit)
+4. Runtime validations performed by resolvers (e.g. during the execution of the GraphQL operation).
+
+Error types 1. and 2. prevent execution of the request in its entirety; they MUST result in
+status code `4xx` or `5xx`, and SHOULD result in status code `400` (Bad Request).
+Types 3. and 4. may still allow partial execution; they MUST result in
+status code `2xx`, and SHOULD result in status code `200` (Okay).
+
+If a client is not allowed to access the schema, the server MUST respond with
+status code `4xx`, SHOULD respond with `401` (Unauthorized) or `403` (Forbidden) as appropriate,
+and MAY respond with status code `404` (Not Found) if security concerns demand it.
```

Comment:
Is this needed (it is already covered in `MUST respond with
status code '4xx'`)?

If you feel it is still needed, then I would prefer that `if security concerns demand it.` is removed (motivations are unenforceable).

Reply:
I shortened that part and just prescribe a `4xx`.